File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-5.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {number} k
     * @return {number}
     */
    kthSmallestWithArray(root: TreeNode | null, k: number): number {
        let results: number[] = [];

        function traverse(rootNode: TreeNode | null) {
            if (rootNode.left) traverse(rootNode.left);
            results.push(rootNode.val);
            if (rootNode.right) traverse(rootNode.right);
        }
        traverse(root);
        return results[k - 1] || null;
    }
    // TC --> O(H + K)--> O(n) worsrt
    // SC --> O(N)
    kthSmallestIterative(root: TreeNode | null, k: number): number {
        let curr=root;
        let stack=[]

        while(curr || stack.length){
            while(curr){
                stack.push(curr);
                curr=curr.left;
            }
            curr=stack.pop();
            k--;
            if(k===0) return curr.val;
            curr=curr.right
        }
    }
    // TC --> O(H + K)--> O(n) worsrt
    // SC --> O(H) stack
        kthSmallest(root: TreeNode | null, k: number): number {
       let count=0;
       function traverse(rootNode:TreeNode|null): number|null{
        if(rootNode===null) return null;
        const left=traverse(rootNode.left);
        // jab tk main extreme left ni phuch jata tb tk count ni bdhauga 
        if(left!==null)return left;
        count++;
        if(count===k) return rootNode.val;
        return traverse(rootNode.right)

       }
       return traverse(root)
    }
    

}

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        maxPath=[root.val]
        def check(node):
            if not node:  return 0
            leftMax=check(node.left)
            rightMax=check(node.right)
            leftMax=max(0,leftMax)
            rightMax=max(0,rightMax)
            maxPath[0]=max(maxPath[0], leftMax+rightMax+node.val)
            return node.val + max(leftMax, rightMax)
        check(root)
        return maxPath[0]
        